connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -837,6 +837,76 @@ class Dataset[T] private[sql] (
     }
   }
 
+  /**
+   * Joins this Dataset returning a `Tuple2` for each pair where `condition` evaluates to true.
+   *
+   * This is similar to the relation `join` function with one important difference in the result
+   * schema. Since `joinWith` preserves objects present on either side of the join, the result
+   * schema is similarly nested into a tuple under the column names `_1` and `_2`.
+   *
+   * This type of join can be useful both for preserving type-safety with the original object
+   * types as well as working with relational data where either side of the join has column names
+   * in common.
+   *
+   * @param other
+   *   Right side of the join.
+   * @param condition
+   *   Join expression.
+   * @param joinType
+   *   Type of join to perform. Default `inner`. Must be one of: `inner`, `cross`, `outer`,
+   *   `full`, `fullouter`,`full_outer`, `left`, `leftouter`, `left_outer`, `right`, `rightouter`,
+   *   `right_outer`.
+   *
+   * @group typedrel
+   * @since 3.5.0
+   */
+  def joinWith[U](other: Dataset[U], condition: Column, joinType: String): Dataset[(T, U)] = {
+    val joinTypeValue = toJoinType(joinType, skipSemiAnti = true)
+    val joinedNullables = joinTypeValue match {
+      case proto.Join.JoinType.JOIN_TYPE_INNER | proto.Join.JoinType.JOIN_TYPE_CROSS =>
+        Seq(false, false)
+      case proto.Join.JoinType.JOIN_TYPE_FULL_OUTER =>
+        Seq(true, true)
+      case proto.Join.JoinType.JOIN_TYPE_LEFT_OUTER =>
+        Seq(false, true)
+      case proto.Join.JoinType.JOIN_TYPE_RIGHT_OUTER =>
+        Seq(true, false)
+      case e =>
+        throw new IllegalArgumentException(s"Unsupported join type `joinType`: $e")
+    }
+
+    val tupleEncoder = ProductEncoder
```

Comment:
This opens up an interesting can of Worms. The problem is basically that we don't know the encoder for a Dataframe. If you joinWith a `DataFrame` (`Dataset[Row]`)  with a `Dataset[E]` you will end up with a `Dataset[(Row, E)]`, the problem is that we don't know the schema for `Row` and we infer this from the schema in `SparkResult`, this does not work in the `joinWith` scenario...

Reply:
The worm can is sealed. I added some extra logic in `SparkResult` to convert the `UnboundRowEncoder`s to a new encoder inferred from the row schema.

Reply:
The can of worms is only closed if the UnboundRowEncoders are only nested in Products/Rows :)... Fine for now though.